### ISSUE TITLE
[apps/v1]: Deployment with `maxUnavailable: 0` had 0 ready pods

### ISSUE THREAD

**Issue by alexanderConstantinescu** [kubernetes/kubernetes#115226]
### What happened?

Let me begin by saying: I am not sure this is a bug or just me lacking insight into how Deployments are rolled out / managed. In any case: I am looking for feedback and some help determining which case this falls under: bug or lack of understanding. 

I am running a cluster on GKE, on which kube-dns is deployed and running with 2 replicas. During an upgrade of the Kubernetes control + data plane from 1.22 to 1.23 a slight outage was observed whereby no kube-dns pods were _ready_. I have attached all the information I could gather (deployment spec / old and new replica set / current rolled out pod spec and state). For the record: GKE manages kube-dns and I don't have access to the Kube control plane logs at that point in time. 

In any case: from the Deployment specifies `maxUnavailable: 0` and  I would imagine to have, at the very least, one replica of kube-dns running and ready at all times. Instead what happened was the following:

```
kube-dns-6df569cd84-jmdhz - pod started: 2023-01-12T10:27:24Z
kube-dns-6df569cd84-jmdhz - dnsmasq started: 2023-01-12T10:29:59Z
kube-dns-6df569cd84-jmdhz - sidecar started: 2023-01-12T10:30:13Z
kube-dns-6df569cd84-jmdhz - kubedns started: 2023-01-12T10:30:14Z
kube-dns-6df569cd84-jmdhz - containersReady: 2023-01-12T10:30:23Z

kube-dns-6df569cd84-pjh6k - pod started: 2023-01-12T10:28:28Z
kube-dns-6df569cd84-pjh6k - kubedns started: 2023-01-12T10:29:32Z
kube-dns-6df569cd84-pjh6k - dnsmasq started: 2023-01-12T10:29:35Z
kube-dns-6df569cd84-pjh6k - sidecar started: 2023-01-12T10:29:37Z
kube-dns-6df569cd84-pjh6k - conainersReady: 2023-01-12T10:29:48Z  
```  
           
This is just a summary of the state reported in the pod files I have.

Given the above it's clear that from `2023-01-12T10:28:28Z -> 2023-01-12T10:29:48Z` no kube-dns pod was ready on the cluster. This outage of 1.5 minutes caused alerts to get triggered on our cluster, reporting problems with DNS resolution - which is normal, given that no kube-dns pod was ready.    
 

### What did you expect to happen?

I would expect to have at least one replica running and ready at all times given the deployment spec.

### How can we reproduce it (as minimally and precisely as possible)?

I don't think this is easily reproducible. I've upgraded multiple clusters from the 1.22 to 1.23, and only a handful displayed this behaviour leading me to believe that this isn't pathologically triggered and might be caused by a race condition or similar. 

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: 4.10.0
Kubernetes Version: v1.23.14-gke.1800
```

</details>

### Cloud provider

<details>
GKE
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by alexanderConstantinescu** [comment]
[kube-dns.zip](https://github.com/kubernetes/kubernetes/files/10465689/kube-dns.zip)

**Comment by alexanderConstantinescu** [comment]
I've attached a zip file with the following content: 

```
-rw-r--r--@  1 aconstantinescu  staff   8702 Jan 19 18:42 kube-dns-6df569cd84-jmdhz
-rw-r--r--@  1 aconstantinescu  staff   8459 Jan 19 18:42 kube-dns-6df569cd84-pjh6k
-rw-r--r--   1 aconstantinescu  staff  10692 Jan 19 18:41 kube-dns.deployment
-rw-r--r--   1 aconstantinescu  staff   6556 Jan 20 11:43 kube-dns.rsnew
-rw-r--r--@  1 aconstantinescu  staff   6433 Jan 20 11:44 kube-dns.rsold
-rw-r--r--   1 aconstantinescu  staff  13104 Jan 20 12:07 kube-dns.zip
```

**Comment by dpasiukevich** [comment]
I'm trying to understand the statement: 

> Given the above it's clear that from 2023-01-12T10:28:28Z -> 2023-01-12T10:29:48Z no kube-dns pod was ready on the cluster. 

There still should be kube-dns pods from previous deployment during that time. So there will be at least 1 kube-dns pod running all the time.

Usually DNS disruptions during cluster upgrade happen as new kube-dns pods need to build cache which decreases performance of kube-dns and may result in DNS resolution errors for some period.

The suggestion to mitigate this - is to enable nodelocaldns. The issue may still happen if both kube-dns and nodelocaldns are upgraded during cluster upgrade however it will be localized to only some nodes since only [10%](https://github.com/kubernetes/kubernetes/blob/2e1da6590e5d40c6417735a1d9b9f4aa67fd02da/cluster/addons/dns/nodelocaldns/nodelocaldns.yaml#L116) will be unavailable at any given time.

**Comment by alexanderConstantinescu** [comment]
> I'm trying to understand the statement:

Yes, you are right that the statement can be viewed as incomplete when judging solely based on the provided pod state. However, we did have alerts firing during that one minute which coincides with the interval: `2023-01-12T10:28:28Z -> 2023-01-12T10:29:48Z`, here's a screenshot of that

<img width="1536" alt="Screenshot 2023-01-23 at 11 11 57" src="https://user-images.githubusercontent.com/17576431/214016670-98d9f111-5786-4b56-a8fe-6f9820b94e53.png">

Is there a possibility to find the pod state for the old replicas that have been reaped?  I don't have access to any control plane logs to be able to spelunk that any further. 

> The issue may still happen if both kube-dns and nodelocaldns are upgraded during cluster upgrade however it will be localized to only some nodes since only [10%](https://github.com/kubernetes/kubernetes/blob/2e1da6590e5d40c6417735a1d9b9f4aa67fd02da/cluster/addons/dns/nodelocaldns/nodelocaldns.yaml#L116) will be unavailable at any given time.

One thing I don't follow though is that kube-dns specifies `maxUnavailable: 0`, if nodelocaldns specifies `10%`, how is that any better?

**Comment by dpasiukevich** [comment]
> One thing I don't follow though is that kube-dns specifies maxUnavailable: 0, if nodelocaldns specifies 10%, how is that any better?

kube-dns has 1 pod per 15 K8S nodes per default autoscaling options.
While nodelocaldns has 1 pod on each K8S node.

So in 30 nodes cluster, there would be 30 nodelocaldns and 2 kube-dns pods.  

And you cannot have 0 kube-dns pods, as it's the DNS provider
While for nodelocaldns `maxUnavailable: 10%` makes sense as the nodelocaldns is just a DNS cache proxy.

**Comment by dpasiukevich** [comment]
As for the DNS failure for 1 minute, as I've said, the problem is likely is that new kube-dns pod was not able to serve DNS queries properly for 1 minute after the start until it built its cache. But there was always at least 1 kube-dns pod.

**Comment by alexanderConstantinescu** [comment]
> As for the DNS failure for 1 minute, as I've said, the problem is likely is that new kube-dns pod was not able to serve DNS queries properly for 1 minute after the start until it built its cache. But there was always at least 1 kube-dns pod.

OK, the DNS issue is beyond the scope of this issue. But for what concerns Deployments: isn't pod readiness taken into account during the rollout sequence? The way I interpret things is:

0. 2 v1 pods running 
[rollout is triggered]
1. 1 v1 pod is stopped 
2. 1 v2 pod is started 
3. second v1 pod is stopped 
4. second v2 pod is started
 
Whereas, I would expect the sequence to be:
 
0. 2 v1 pods running 
[rollout is triggered]
1. 1 v1 pod is stopped 
2. 1 v2 pod is started 
3. 1 v2 pod is ready
4. second v1 pod is stopped 
5. second v2 pod is started
6. second v2 pod is ready

Without taking readiness into account, doesn't the rollout sequence allow for the "availability SLA" to be broken?
